hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddedForeignKeyDescriptor.java

```diff
@@ -115,35 +116,39 @@ public DomainResult createCollectionFetchDomainResult(
 			final TableReference tableReference = tableGroup.resolveTableReference( keyColumnContainingTable );
 			final String identificationVariable = tableReference.getIdentificationVariable();
 
-			List<SqlSelection> sqlSelections = new ArrayList<>();
-			for ( int i = 0; i < keyColumnExpressions.size(); i++ ) {
-				final JdbcMapping jdbcMapping = jdbcMappings.get( i );
-				final String columnExpression = targetColumnExpressions.get( i );
-				final SqlSelection sqlSelection = sqlExpressionResolver.resolveSqlSelection(
-						sqlExpressionResolver.resolveSqlExpression(
-								SqlExpressionResolver.createColumnReferenceKey(
-										tableReference,
-										columnExpression
+			List<SqlSelection> sqlSelections = new ArrayList<>( targetColumnMappings.getColumnSpan() );
+			targetColumnMappings.forEachColumn(
```

Comment:
Another case where this code would be a lot cleaner passing along the `ColumnMapping`